src/solver.rs

```diff
@@ -0,0 +1,134 @@
+            v.as_slice()
+                .try_into()
+                .ok()
+                .map(|hash| ScriptKind::ScriptHash { hash })
+        }
+
+        // Pay-to-Public-Key-Hash (P2PKH)
+        [op::DUP, op::HASH160, Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::EQUALVERIFY, op::CHECKSIG] => {
+            v.as_slice()
+                .try_into()
+                .ok()
+                .map(|hash| ScriptKind::PubKeyHash { hash })
+        }
+
+        // Provably prunable, data-carrying output.
+        //
+        // So long as the script passes the `is_unspendable()` test, and all but the first
+        // byte passes the `is_push_only()` test, we don't care what exactly is in the
+        // script.
+        [op::RETURN, rest @ ..] => rest
+            .iter()
+            .map(|op| op.extract_push_value().ok().cloned())
+            // This `.collect()` is equivalent to `script::Opcode::is_push_only` for the
+            // subset of inputs consisting of entirely valid `PushValue`s, which is all we
+            // care about here.
+            .collect::<Option<_>>()
+            .map(|data| ScriptKind::NullData { data }),
+
+        // Legacy Pay-to-Public-Key (P2PK), which only occurred in early Zcash coinbase outputs.
+        [Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::CHECKSIG] => {
+            (v.len() == PubKey::SIZE || v.len() == PubKey::COMPRESSED_SIZE).then(|| {
+                ScriptKind::PubKey {
+                    data: v.clone().to_vec().try_into().expect("bounds checked"),
+                }
+            })
+        }
+
+        // Pay-to-Multi-Signature (P2MS)
+        [Opcode::PushValue(PushValue::SmallValue(required)), pubkeys @ .., Opcode::PushValue(PushValue::SmallValue(keys)), op::CHECKMULTISIG] =>
+        {
+            match (u8::try_from(required.to_num()), u8::try_from(keys.to_num())) {
+                (Ok(required @ 0..=16), Ok(keys @ 0..=16)) => {
+                    // The remaining opcodes must be `PushData`s that are valid lengths for pubkeys.
+                    let pubkeys = pubkeys
+                        .iter()
+                        .map(|op| match op {
+                            Opcode::PushValue(data) => {
+                                let pubkey_bytes = data.value();
+                                // Equivalent to `CPubKey::GetLen`
+                                let expected_len = match pubkey_bytes.first() {
+                                    Some(2 | 3) => PubKey::COMPRESSED_SIZE,
+                                    Some(4 | 6 | 7) => PubKey::SIZE,
+                                    _ => 0,
+                                };
+                                // Equivalent to `CPubKey::ValidSize`
+                                (!pubkey_bytes.is_empty() && expected_len == pubkey_bytes.len())
+                                    .then(|| pubkey_bytes.try_into().expect("bounds checked"))
```

Comment:
It’d be nice to avoid the lurking bug here – the `!pubkey_bytes.is_empty()` is only to protect against the `0` that represents an invalid result in the `match`.
```suggestion
                                // Equivalent to `CPubKey::GetLen`
                                let expected_len = match pubkey_bytes.first() {
                                    Some(2 | 3) => Some(PubKey::COMPRESSED_SIZE),
                                    Some(4 | 6 | 7) => Some(PubKey::SIZE),
                                    _ => None,
                                };
                                // Equivalent to `CPubKey::ValidSize`
                                expected_len
                                    .filter(|expected| expected == pubkey_bytes.len())
                                    .map(|_| pubkey_bytes.try_into().expect("bounds checked"))
```
or
```rust
                                (expected_len == Some(pubkey_bytes.len()))
                                    .then(|| pubkey_bytes.try_into().expect("bounds checked"))
```

Reply:
Indeed, and I kept the `!pubkey_bytes.is_empty()` intentionally to match the C++ code for easier review. I'll add a follow-on commit that does the oxidation.